lib/puppet/parser/scope.rb

```diff
@@ -738,17 +815,27 @@ def to_s
 
   alias_method :inspect, :to_s
 
-  # remove ephemeral scope up to level
-  # TODO: Who uses :all ? Remove ??
+  # Pop ephemeral scopes up to level and return them
   #
+  # @param level [Symbol,Fixnum] The symbol `:all` or a positive integer
+  # @return [Array] the removed ephemeral scopes
+  # @api private
   def unset_ephemeral_var(level=:all)
     if level == :all
+      result = @ephemeral
       @ephemeral = [ MatchScope.new(@symtable, nil)]
+      result
```

Comment:
as per the comment above - don't change this - if you want to improve this, add a deprecation warning; there should be no calls to unset_ephemeral_var(:all) and if someone does that they should be told. The proposed change would just break them (or make them screw up as it changes the broken API. (And yes, this is horrible).